### ISSUE TITLE
More than 3 pixels values predicted while training models

### ISSUE THREAD

**Issue by vasudev-sharma** [axondeepseg/axondeepseg#424]
As mentioned in the last meeting, I was investigating transfer learning on Simeon's data. For that, I used pretrained weights of our `default_TEM_model` hosted in OSF as our starting point. So, the idea was that we want to fine tune our model on Simeon's data. 

Here are the metrics which are compared against our default TEM model
![image](https://user-images.githubusercontent.com/31419210/102479488-f8440480-4084-11eb-936a-333a5e04fe58.png)

**NOTE:**
 - Orange legend depicts our **default TEM model**
  - Silver legend depicts the **fine tuned model** on Simeon's data

So, you can clearly see the validation metrics of the fine tuned model is good in comparison to our default model except our val_dice_myelin score.

Seeing from the metrics we expect the results of the fine tuned model to be alteast better or worst case scenario-- be equivalent.

Look at the results below on **image 2** of Simeon's nerve segmentation dataset.

**1. default TEM model**
![image](https://user-images.githubusercontent.com/31419210/102481111-4b1ebb80-4087-11eb-8d86-96e9ed3cb52c.png)

**2. Fine tuned TEM model**
![image](https://user-images.githubusercontent.com/31419210/102481564-e31ca500-4087-11eb-9a3b-e14d65fd30de.png)

Looking at the prediction of the fine tuned TEM model, it seems weird in the sense that although we're getting good validation metrics for this model, the prediction results are bad ( I can't explain this prediction result for the fine tuned model )

To investigate this, I visualized how the predictions on validation set change while training the model. 

Here is the gif image of masks predicted ( starting from our `pretrained_TEM_model`and fine tuning this model) by our network :-

![prediction_model_100_epochs](https://user-images.githubusercontent.com/31419210/102494608-eec59700-409a-11eb-952d-f9a61255212e.gif)

So the GIF contains
1. mask predicted by our **default TEM model/ pretrained TEM model**
2. mask predicted when model is trained for **1 epoch** 
3. mask predicted when model is trained for **20 epochs**
4. mask predicted when model is trained for **40 epochs**
5. mask predicted when model is trained for **80 epochs**
6. mask predicted when model is trained for **100 epochs**

If you zoom in the masks predicted, you can clearly see that initially the model only predicted 3 pixel values [ 0, 127, 255], but as the training progress the model predicts more than 3 classes. I suspect this could be the reasons I got weird prediction results on Simeon's data even though the validation metrics were good enough. 

**Zoomed in mask**
1. default TEM model / pretrained TEM model
![image](https://user-images.githubusercontent.com/31419210/102497657-eec79600-409e-11eb-958e-2eaf01f3064f.png)

2. mask predicted when model is trained for 1 epoch 
![image](https://user-images.githubusercontent.com/31419210/102497943-5382f080-409f-11eb-8303-12a3fdf82e4f.png)

3. mask predicted when model is trained for 100 epochs
![image](https://user-images.githubusercontent.com/31419210/102497757-0e5ebe80-409f-11eb-8dd9-922e3a998413.png)

PS: I noticed the GIF is not rendering here, so please click on the GIF image to see how the model's prediction change when model is trained. Also, I'll be adding more details about this issue.

**Comment by vs74** [comment]
I know I've added a lot of information here, so if anything is unclear let me know. This issue is related to https://github.com/neuropoly/axondeepseg/issues/139. Intially, I thought it could be to our gold standards / manual masks, but I double checked all the gold standards contain 3 pixel intensities only [0, 127, 255].  . One thing I can't seem to understand till now is that why this issue is not encountered with our `default_SEM` and `default_TEM model `.



**Note:-** I kept all the code exactly the same except one modification I did, that is, I initialized the model's weight with our `default TEM model` to fine tune on Simeon's data.

**Comment by vs74** [comment]
Ahaa!! A great article to look at  https://distill.pub/2016/deconv-checkerboard/. This effect is known as **checkerbaord artifact**, which is encountered due to deconvolution layers of a model and supposedly I think that the [expansion phase](https://github.com/neuropoly/axondeepseg/blob/3862aa223bfe1710ba4dd68965df8b131c503755/AxonDeepSeg/network_construction.py#L118-L143) of our network could be a cause of this. I am curious to know if IVADOMED team has encountered this issue in their multiclass segmentation models.

**Comment by mathieuboudreau** [comment]
> I know I've added a lot of information here, so if anything is unclear, let me know. This issue is related to #139. Intially, I thought it could due to our gold standards / manual masks, but I double checked all the gold standards and they contain 3 pixel intensities only [0, 127, 255]. One thing I can't seem to understand till now is that why this issue is not encountered with our `default_SEM` and `default_TEM model `.
> 
> **Note:-** I kept all the code exactly the same except one modification I did, that is, I initialized the model's weight with our `default TEM model` to fine tune on Simeon's data.

Interesting. We actually had resolved #139, the problem was that some of our resampling/transformations/data augmentations were using default function arguments, which was setting interpolation to be linear instead of nearest-neighbour like we need for our 3-class images.

**Comment by mathieuboudreau** [comment]
Stale - no longer using this model/code
